core/src/main/java/io/questdb/griffin/engine/functions/str/SubStringFunctionFactory.java

```diff
@@ -0,0 +1,162 @@
+            return StrConstant.NULL;
+        }
+        if (lenFunc.isConstant()) {
+            int len = lenFunc.getInt(null);
+            if (len < 0) {
+                throw SqlException.$(position, "negative substring length is not allowed");
+            } else if (len == 0) {
+                return StrConstant.EMPTY;
+            }
+        }
+        return new SubStringFunc(strFunc, startFunc, lenFunc);
+    }
+
+    private static class SubStringFunc extends StrFunction implements TernaryFunction {
+
+        private final Function strFunc;
+        private final Function startFunc;
+        private final Function lenFunc;
+
+        private final StringSink sinkA = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final boolean isSimplifiable;
+
+        public SubStringFunc(Function strFunc, Function startFunc, Function lenFunc) {
+            this.strFunc = strFunc;
+            this.startFunc = startFunc;
+            this.lenFunc = lenFunc;
+
+            this.isSimplifiable = startFunc.isConstant() && lenFunc.isConstant()
+                    && startFunc.getInt(null) + lenFunc.getInt(null) < 1;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return startFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return lenFunc;
+        }
+
+        @Override
+        public CharSequence getStr(Record rec) {
+            return getStr0(rec, sinkA);
+        }
+
+        @Override
+        public CharSequence getStrB(final Record rec) {
+            return getStr0(rec, sinkB);
+        }
+
+        @Override
+        public int getStrLen(final Record rec) {
+            int strLen = strFunc.getStrLen(rec);
+            int rawStart = startFunc.getInt(rec);
+            int len = lenFunc.getInt(rec);
+            if (strLen == TableUtils.NULL_LEN
+                    || rawStart == Numbers.INT_NaN
+                    || len == Numbers.INT_NaN) {
+                return TableUtils.NULL_LEN;
+            }
+
+            int start = Math.max(0, rawStart - 1);
+            if (len == 0
+                    || start > strLen
+                    || rawStart + len < 1) {
+                return 0;
+            }
+            int end = Math.min(strLen, rawStart + len - 1);
+            return Math.max(0, end - start);
+        }
+
+        @Nullable
+        private StringSink getStr0(Record rec, StringSink sink) {
+            CharSequence str = strFunc.getStr(rec);
+            if (str == null || isSimplifiable) {
+                return null;
+            }
+            int rawStart = startFunc.getInt(rec);
+            int len = lenFunc.getInt(rec);
+            if (rawStart == Numbers.INT_NaN || len == Numbers.INT_NaN) {
+                return null;
+            }
```

Comment:
This would be the first time function throws RuntimeException so you'd need to test it thouroughly .

Reply:
If so, things would be inconsistent with `const len < 0`, whic would raise a SqlException, as https://github.com/questdb/questdb/pull/2251#discussion_r917896421 mentioned.

Maybe we could make both scenarios return null?

Reply:
ok, let's stick to exception